File: leetcode/graph/3383_Minimum_Runes_to_Add_to_Cast_Spell/codes/kosaraju.py
from collections import defaultdict
from typing import List

class Solution:
    def minRunesToAdd(self, n: int, crystals: List[int], flowFrom: List[int], flowTo: List[int]) -> int:
        adj = defaultdict(list)
        transpose = defaultdict(list)
        for u, v in zip(flowFrom, flowTo):
            adj[u].append(v)
            transpose[v].append(u)
        
        visited = set()
        stack = []
        
        def dfs(u):
            visited.add(u)
            for v in adj[u]:
                if v not in visited:
                    dfs(v)
            stack.append(u)
        
        def dfs_transpose(u, scc):
            visited.add(u)
            scc.append(u)
            for v in transpose[u]:
                if v not in visited:
                    dfs_transpose(v, scc)
        
        for i in range(n):
            if i not in visited:
                dfs(i)
        
        visited.clear()
        sccs = []
        while stack:
            node = stack.pop()
            if node not in visited:
                scc = []
                dfs_transpose(node, scc)
                sccs.append(scc)
        
        scc_map = {}
        for i, scc in enumerate(sccs):
            for node in scc:
                scc_map[node] = i
        
        dag_indegree = [0]*len(sccs)
        edges_set = set()
        for u in range(n):
            for v in adj[u]:
                if scc_map[u] != scc_map[v]:
                    if (scc_map[u], scc_map[v]) not in edges_set:
                        edges_set.add((scc_map[u], scc_map[v]))
                        dag_indegree[scc_map[v]] += 1
        
        crystal_scc = set(scc_map[c] for c in crystals)
        
        count = 0
        for i in range(len(sccs)):
            if dag_indegree[i] == 0 and i not in crystal_scc:
                count += 1
        
        return count

File: leetcode/graph/3383_Minimum_Runes_to_Add_to_Cast_Spell/codes/kosaraju.cpp
#include <vector>
#include <stack>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int minRunesToAdd(int n, vector<int>& crystals, vector<int>& flowFrom, vector<int>& flowTo) {
        vector<vector<int>> adj(n), transpose(n);
        for (int i = 0; i < (int)flowFrom.size(); i++) {
            adj[flowFrom[i]].push_back(flowTo[i]);
            transpose[flowTo[i]].push_back(flowFrom[i]);
        }
        
        vector<bool> visited(n, false);
        stack<int> st;
        for (int i = 0; i < n; i++) if (!visited[i]) dfs(i, adj, visited, st);
        
        fill(visited.begin(), visited.end(), false);
        vector<vector<int>> sccs;
        while (!st.empty()) {
            int node = st.top(); st.pop();
            if (!visited[node]) {
                vector<int> scc;
                dfsTranspose(node, transpose, visited, scc);
                sccs.push_back(move(scc));
            }
        }
        
        vector<int> sccMap(n);
        for (int i = 0; i < (int)sccs.size(); i++) {
            for (int node : sccs[i]) sccMap[node] = i;
        }
        
        vector<int> indegree((int)sccs.size(), 0);
        unordered_set<long long> edgeSet;
        for (int u = 0; u < n; u++) {
            for (int v : adj[u]) {
                if (sccMap[u] != sccMap[v]) {
                    long long code = ((long long)sccMap[u] << 32) | sccMap[v];
                    if (edgeSet.find(code) == edgeSet.end()) {
                        edgeSet.insert(code);
                        indegree[sccMap[v]]++;
                    }
                }
            }
        }
        
        unordered_set<int> crystalScc;
        for (int c : crystals) crystalScc.insert(sccMap[c]);
        
        int count = 0;
        for (int i = 0; i < (int)sccs.size(); i++) {
            if (indegree[i] == 0 && crystalScc.count(i) == 0) count++;
        }
        return count;
    }
    
private:
    void dfs(int u, vector<vector<int>>& adj, vector<bool>& visited, stack<int>& st) {
        visited[u] = true;
        for (int v : adj[u]) {
            if (!visited[v]) dfs(v, adj, visited, st);
        }
        st.push(u);
    }
    
    void dfsTranspose(int u, vector<vector<int>>& transpose, vector<bool>& visited, vector<int>& scc) {
        visited[u] = true;
        scc.push_back(u);
        for (int v : transpose[u]) {
            if (!visited[v]) dfsTranspose(v, transpose, visited, scc);
        }
    }
};
